server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -480,17 +482,38 @@ public void deleteSnapshot(SnapshotId snapshotId, long repositoryStateId) {
                     // snapshotting indices of the same name
                     logger.debug(() -> new ParameterizedMessage("[{}] index [{}] no longer part of any snapshots in the repository, " +
                         "but failed to clean up its index folder due to the directory not being empty.", metadata.name(), indexId), dnee);
+                    deleteSuccess = false;
                 } catch (IOException ioe) {
                     // a different IOException occurred while trying to delete - will just log the issue for now
                     logger.debug(() -> new ParameterizedMessage("[{}] index [{}] no longer part of any snapshots in the repository, " +
                         "but failed to clean up its index folder.", metadata.name(), indexId), ioe);
+                    deleteSuccess = false;
+                }
+                if (deleteSuccess) {
+                    indicesBlobContainer.deleteBlob(tombstoneBlob(indexId));
```

Comment:
One alternative here could be to not delete these ever:
* The storage cost of keeping them around would be minimal
* It would solve the issue of "rogue" data nodes writing shard data for an already deleted snapshot and us missing this situation because they do so after this code finishes